### PR TITLE
Support ConfigMap synchronization function for nacos

### PR DESCRIPTION
**Pull request by wukong121**
## The purpose of the change

Support ConfigMap synchronization function for nacos

## Brief changelog

update ConfigMapSyncTask
update KubernetesConfigMapSyncServer
add ConfigMapSyncConfig
add kubeconfig.yaml

### CODE REVIEW

File: distribution/conf/application.properties [hunk 1]
```diff
@@ -33,16 +33,16 @@ server.port=8848
 #*************** Config Module Related Configurations ***************#
 ### If use MySQL as datasource:
 ### Deprecated configuration property, it is recommended to use `spring.sql.init.platform` replaced.
-# spring.datasource.platform=mysql
-# spring.sql.init.platform=mysql
+ spring.datasource.platform=mysql
```
Comment (slievrly) [inline distribution/conf/application.properties:36]: revert  this file change.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 2]
```diff
@@ -43,17 +45,18 @@ public class ConfigMapSyncTask {
...
+        apiClient.setHttpClient(httpClient);
+        nacosConfigService = NacosFactory.createConfigService(apiClient.getBasePath());
     }
     
-    // 每小时从k8s中list所有的configmaps, 如果nacos中没有则添加
-    @Scheduled(fixedDelay = 3600000) // 每小时对帐
+    /**
+     * Checks and synchronizes configuration maps.
+     */
+    @Scheduled(fixedDelay = 3600000)
```
Comment (stone-98) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:59]: 我有几个疑问哈：
1、定期对账为什么要用Spring里面的定时任务，感觉用JDK里面的定时任务会更好，还可以定义一个配置项就是对账的间隔时间。
2、这个定期对账是通过dataId+group去查configService，并没有校验内容，这样是否会有问题呢？例如一种情况，当Nacos第一次启动，Nacos已经把ConfigMap加载到db，那么当Nacos宕机时，在这个期间ConfigMap发生了变化，此时Nacos无法感知，当再次启动这个定时对账任务也无法感知。
3、还有一种就是Nacos如何保证最新版本呢？比如Nacos首先对configMap进行watch，正好对账任务发现此时Nacos没有，就进行发布配置，这时候configMap正好发生变化，也进行发布配置，此时nacos如何保证db存储的是最新的配置呢，k8s的资源的metadata中有resourceVersion字段，nacos是否应该对它进行存储从而进行保证呢？

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncConfig.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @author wangyixing
+ */
+@Component
+public class ConfigMapSyncConfig {
+    
+    @Value("${nacos.k8s.configMap.enabled:false}")
+    private boolean enabled;
+    
+    @Value("${nacos.k8s.configMap.responsible:false}")
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncConfig.java:35]: unused configuration.
What is the meaning of the configuration item?

File: distribution/logs/start.out [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/Users/peter/Library/Java/JavaVirtualMachines/openjdk-19.0.2/Contents/Home/bin/java   -Xms512m -Xmx512m -Xmn256m -Dnacos.standalone=true -Dnacos.member.list= -Xlog:gc*:file=/Users/peter/Documents/nacos/distribution/logs/nacos_gc.log:time,tags:filecount=10,filesize=100m -Dloader.path=/Users/peter/Documents/nacos/distribution/plugins,/Users/peter/Documents/nacos/distribution/plugins/health,/Users/peter/Documents/nacos/distribution/plugins/cmdb,/Users/peter/Documents/nacos/distribution/plugins/selector -Dnacos.home=/Users/peter/Documents/nacos/distribution -jar /Users/peter/Documents/nacos/distribution/target/nacos-server.jar  --spring.config.additional-location=file:/Users/peter/Documents/nacos/distribution/conf/ --logging.config=/Users/peter/Documents/nacos/distribution/conf/nacos-logback.xml --server.max-http-header-size=524288
```
Comment (slievrly) [inline distribution/logs/start.out:1]: delete this file

File: persistence/src/test/java/com/alibaba/nacos/persistence/configuration/condition/ConfigChangeConfigsTest.java [hunk 5]
```diff
@@ -1,98 +0,0 @@
...
-import org.springframework.mock.env.MockEnvironment;
-
-import java.util.Properties;
-
-import static org.junit.Assert.assertEquals;
-
-public class ConfigChangeConfigsTest {
-
-    private static final boolean TEST_WEBHOOK_ENABLED = false;
-
```
Comment (slievrly) [inline persistence/src/test/java/com/alibaba/nacos/persistence/configuration/condition/ConfigChangeConfigsTest.java:33]: why remove the UT case ?

File: pom.xml [hunk 6]
```diff
@@ -153,7 +153,7 @@
         <HikariCP.version>3.4.2</HikariCP.version>
         <jraft-core.version>1.3.12</jraft-core.version>
         <rpc-grpc-impl.version>${jraft-core.version}</rpc-grpc-impl.version>
-        <SnakeYaml.version>2.0</SnakeYaml.version>
+        <SnakeYaml.version>1.33</SnakeYaml.version>
```
Comment (slievrly) [inline pom.xml:156]: What are the reasons for using the lower version?

Comment (wukong121) [reply pom.xml:156]: I got a bug and found solution by the following link:
https://stackoverflow.com/questions/70154082/getting-java-lang-nosuchmethoderror-org-yaml-snakeyaml-yaml-init-while-runnin
bugs:
```bash
Caused by: java.lang.NoSuchMethodError: org.yaml.snakeyaml.constructor.SafeConstructor: method <init>()V not found
```

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncConfig.java (hunk 3 above)
Comment (wukong121) [reply config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncConfig.java:35]: `enabled` is the on-off switch of configMap synchronization job.
`responsible` is designed to decide which node is responsible for data persistence job.But we did not have time to develop this part, so I leave the config for future TODO.

File: persistence/src/test/java/com/alibaba/nacos/persistence/configuration/condition/ConfigChangeConfigsTest.java (hunk 5 above)
Comment (wukong121) [reply persistence/src/test/java/com/alibaba/nacos/persistence/configuration/condition/ConfigChangeConfigsTest.java:33]: Yeah, it's a truely confusion. I don't know why I can not run this UT properly. I can not build the whole job with this UT. So I deleted it :(

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 7]
```diff
@@ -18,60 +18,133 @@
...
-        coreV1Api = new CoreV1Api();
-        
-        nacosConfigService = NacosFactory.createConfigService("TODO");
+    /**
+     * Initialize ConfigMapSyncTask.
+     * @throws IOException getOutsideApiClient failed
+     */
+    @PostConstruct
+    public void init() throws IOException {
+        if (configMapSyncConfig.isOutsideCluster()) {
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:68]: The logical coupling built here is too heavy, and a layer of Factory should be extracted for both inside and outside forms of APIClient.

---

此处构建的逻辑耦合过重，对于inside和outside两种形态的APIClient，应该提取出一层Factory。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 8]
```diff
@@ -18,60 +18,133 @@
...
+                    configMapSyncService.publishConfigMap(configMap, apiClient.getBasePath());
+                } else {
+                    if (!compareContent(content, configInfo.getContent())) {
+                        Loggers.MAIN.info("[{}] find content difference.", "configMap-sync");
+                        configMapSyncService.publishConfigMap(configMap, apiClient.getBasePath());
+                    }
                 }
             }
         } catch (Exception e) {
             e.printStackTrace();
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:104]: log is required.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 9]
```diff
@@ -18,60 +18,133 @@
...
+     * Checks and synchronizes configuration maps.
+     */
+    @Scheduled(fixedDelay = 3600000)
     public void checkAndSyncConfigMaps() {
         try {
-            V1ConfigMapList configMapList = (V1ConfigMapList) coreV1Api.listConfigMapForAllNamespacesCall(null, null,
-                    null, null, null, null, null, null, null, null, null);
-            for (V1ConfigMap configMap : configMapList.getItems()) {
-                String dataId = configMap.getMetadata().getName();
+            List<V1ConfigMap> configMapList = coreV1Api.listConfigMapForAllNamespaces(null, null, null, null, null,
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:85]: It is recommended that you do not assign Boolean and Integer to null directly. You need to specify a clear service meaning or default value. The limit value here needs to consider the appropriate value.

---

建议不要将Boolean和Integer直接赋值为null，需要指定明确业务含义或默认值。这里的limit值需要考虑下合适的值。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 10]
```diff
@@ -18,60 +18,133 @@
...
+                    null, null, null, null, null).getItems();
+            for (V1ConfigMap configMap : configMapList) {
+                String dataId = Objects.requireNonNull(configMap.getMetadata()).getName();
                 String group = "K8S_GROUP";
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:89]: ConfigMapSyncConfig.K8S_GROUP

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java (hunk 8 above)
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:104]: The exception here needs to be considered whether to stop the exception directly or skip the current dataId synchronization and proceed to the next one, I personally prefer the latter.

---

这里异常需要考虑是出现异常直接停止还是跳过当前dataId同步继续下一个，我个人倾向于后者。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 11]
```diff
@@ -18,60 +18,133 @@
...
+    }
+    
+    /**
+     * Convert and compare configMap and nacos content.
+     * @param configMapContent configMap's content
+     * @param nacosContent nacos's content
+     * @return whether contents are equal.
+     */
+    private Boolean compareContent(String configMapContent, String nacosContent) {
+        configMapContent = configMapContent.substring(1, configMapContent.length() - 1);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:115]: What is returned here is not necessarily json style, there may be yaml format.
image


---

这里不一定返回的是json样式，可能存在yaml格式

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java [hunk 12]
```diff
@@ -18,60 +18,133 @@
...
+        Map<String, String> nacosPair = contentToMap(nacosContent);
+        return configMapPair.equals(nacosPair);
+    }
+    
+    private Map<String, String> contentToMap(String content) {
+        String[] pairs = content.split(", ");
+        Map<String, String> hashMap = new HashMap<>(pairs.length);
+        String currentKey = null;
+        for (String pair : pairs) {
+            if (isPair(pair)) {
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncTask.java:127]: need  consider the yaml format

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+        ParamUtils.checkParam(dataId, group, "datumId", content);
+        
+        configForm.setDataId(dataId);
+        configForm.setGroup(group);
+        configForm.setNamespaceId(namespaceId);
+        configForm.setContent(content);
+        configForm.setSrcUser(ConfigMapSyncConfig.SRC_USER);
+        
+        ConfigRequestInfo configRequestInfo = new ConfigRequestInfo();
+        configRequestInfo.setSrcIp(truncateUrl(srcIp));
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:288]: Why add the protocol type here and try to avoid returning localhost.

---

为什么此处要添加协议类型，应当尽量避免返回localhost.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+ * ConfigMap's synchronization server.
+ *
+ * @author wangyixing
+ */
+@Service
+public class ConfigMapSyncServiceImpl implements ConfigMapSyncService {
+    
+    private Gson gson = new Gson();
+    
+    private boolean isRunning;
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:72]: not used.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+    
+    @Autowired
+    private ConfigOperationService configOperationService;
+    
+    @Autowired
+    private NamespaceOperationService namespaceOperationService;
+    
+    @Autowired
+    @Qualifier("embeddedConfigInfoPersistServiceImpl")
+    private ConfigInfoPersistService configInfoPersistService;
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:87]: It is better to build with Factory rather than hard-coded injection by name.

---

最好是使用Factory构建，而不是按照名字硬编码注入。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+                            params.resourceVersion, null, params.timeoutSeconds, params.watch, null);
+                }, V1ConfigMap.class, V1ConfigMapList.class);
+        configInformer.addEventHandler(new ResourceEventHandler<V1ConfigMap>() {
+            @Override
+            public void onAdd(V1ConfigMap obj) {
+                if (obj == null || obj.getMetadata() == null || obj.getData() == null) {
+                    return;
+                }
+                Loggers.MAIN.info("[{}] adding configMap...", "configMap-sync");
+                ConfigInfo configInfo = configMapToNacosConfigInfo(obj);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:152]: not used.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+                String srcIp = apiClient.getBasePath();
+                
+                try {
+                    publishConfigMap(obj, srcIp);
+                } catch (NacosException e) {
+                    throw new RuntimeException(e);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:158]: Here I prefer error logging rather than throwing exceptions directly, considering the robustness of thread running.

----

这里我倾向于error日志而不是直接throw异常，考虑线程运行的健壮性。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+        });
+        // 启动SharedInformerFactory在后台运行事件监听器
+        factory.startAllRegisteredInformers();
+    }
+    
+    private void compareConfigMaps(V1ConfigMap previousConfigMap, V1ConfigMap currentConfigMap) {
+        // 将先前的 ConfigMap 和当前的 ConfigMap 转换成 JSON 字符串
+        String previousJson = gson.toJson(previousConfigMap);
+        String currentJson = gson.toJson(currentConfigMap);
+
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:196]: Is serialization required here? V1ConfigMap.data contains data in Map format.

---

这里序列化是否是必需的？V1ConfigMap.data包含了Map格式的数据。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+                }
+            }
+            
+            @Override
+            public void onUpdate(V1ConfigMap oldObj, V1ConfigMap newObj) {
+                Loggers.MAIN.info(
+                        "[{}] update configMap " + oldObj.getMetadata().getName() + " to " + newObj.getMetadata()
+                                .getName(), "configMap-sync");
+                compareConfigMaps(oldObj, newObj);
+                ConfigInfo configInfo = configMapToNacosConfigInfo(newObj);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:168]: not used.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+                String srcIp = apiClient.getBasePath();
+                try {
+                    publishConfigMap(newObj, srcIp);
+                } catch (NacosException e) {
+                    throw new RuntimeException(e);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:173]: same above.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+                if (obj == null || obj.getMetadata() == null || obj.getData() == null) {
+                    return;
+                }
+                Loggers.MAIN.info("[{}] delete configMap " + obj.getMetadata().getName(), "configMap-sync");
+                String srcIp = apiClient.getBasePath();
+                deleteConfigMap(obj, srcIp);
+            }
+            
+        });
+        // 启动SharedInformerFactory在后台运行事件监听器
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:188]: comment into En.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+    }
+    
+    /**
+     * publish ConfigMap.
+     * @param configMap k8sConfigMap
+     * @param srcIp source Ip
+     * @throws NacosException nacos exception
+     */
+    public void publishConfigMap(V1ConfigMap configMap, String srcIp) throws NacosException {
+        Loggers.MAIN.info("[{}] Converting configMap to nacos ConfigForm...", "configMap-sync");
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:259]: The add and update operations are distinguished in log printing.

---

日志打印中区分add和update操作。

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+        
+        final ConfigForm configForm = new ConfigForm();
+        String configMapNamespace = configMap.getMetadata().getNamespace();
+        List<Namespace> namespaceList = namespaceOperationService.getNamespaceList();
+        String randomNamespace = UUID.randomUUID().toString();
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:264]: remove below `!matchingNamespace.isPresent()`

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+        Optional<Namespace> matchingNamespace = namespaceList.stream()
+                .filter(namespace -> Objects.equals(namespace.getNamespaceShowName(), configMapNamespace)).findFirst();
+        if (!matchingNamespace.isPresent()) {
+            namespaceOperationService.createNamespace(randomNamespace, configMapNamespace, "");
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:268]: return value check is required.

File: config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+        }
+        String namespaceId = matchingNamespace.map(Namespace::getNamespace).orElse(randomNamespace);
+        String dataId = configMap.getMetadata().getName();
+        String content = getContent(configMap.getData());
+        String group = ConfigMapSyncConfig.K8S_GROUP;
+        Pair<String, String> pair = EncryptionHandler.encryptHandler(dataId, content);
```
Comment (slievrly) [inline config/src/main/java/com/alibaba/nacos/config/server/service/kubernetes/ConfigMapSyncServiceImpl.java:274]: There's no encryption here.

**Review by slievrly — COMMENTED** [review]
Need to be listed `kubernetes.client.openapi `and K8S cluster version of the mapping relationship.

**Comment by stone-98** [comment]
大佬们，我有一个疑问，如果在对账过程中，刚刚好出现configMap发送变化，这样会不会导致老的configMap覆盖掉新的configMap呢？从而导致k8s和nacos中的数据不一致问题？

File: distribution/conf/application.properties [hunk 26]
```diff
@@ -33,16 +33,16 @@ server.port=8848
...
-# db.num=1
+ db.num=1
 
 ### Connect URL of DB:
-# db.url.0=jdbc:mysql://127.0.0.1:3306/nacos?characterEncoding=utf8&connectTimeout=1000&socketTimeout=3000&autoReconnect=true&useUnicode=true&useSSL=false&serverTimezone=UTC
-# db.user.0=nacos
-# db.password.0=nacos
+ db.url.0=jdbc:mysql://127.0.0.1:3306/nacos?characterEncoding=utf8&connectTimeout=1000&socketTimeout=3000&autoReconnect=true&useUnicode=true&useSSL=false&serverTimezone=UTC&allowPublicKeyRetrieval=true
+ db.user=root
+ db.password=wobushiren1!
```
Comment (realJackSun) [inline distribution/conf/application.properties:45]: Unnecessary change

**Review by realJackSun — CHANGES_REQUESTED** [review]

**Review by realJackSun — APPROVED** [review]
